File: examples/experiments/multilang/python/json_dumps.py
import json


def _int_param(params, name, default, minimum, maximum):
    try:
        value = int(params.get(name, default))
    except (TypeError, ValueError, AttributeError):
        value = default
    return max(minimum, min(maximum, value))


def handler(params, context):
    params = params or {}
    records = _int_param(params, "records", 12000, 1000, 50000)
    rounds = _int_param(params, "rounds", 20, 1, 100)

    payload = [
        {
            "id": i,
            "group": i % 97,
            "active": (i % 3) != 0,
            "score": ((i * 2654435761) & 0xFFFFFFFF) / 4294967296.0,
            "name": f"record-{i:06d}",
            "tags": [i % 11, i % 17, i % 23],
        }
        for i in range(records)
    ]

    checksum = 0
    encoded_bytes = 0

    for _ in range(rounds):
        encoded = json.dumps(
            payload,
            separators=(",", ":"),
            sort_keys=True,
        )
        decoded = json.loads(encoded)

        encoded_bytes += len(encoded)
        checksum = (
                           checksum
                           + decoded[-1]["id"]
                           + decoded[records // 2]["group"]
                   ) & 0xFFFFFFFF

    return {
        "benchmark": "json_dumps",
        "records": records,
        "rounds": rounds,
        "encoded_bytes": encoded_bytes,
        "checksum": checksum,
    }

File: examples/experiments/multilang/nodejs/json_dumps.js
function intParam(params, name, defaultValue, min, max) {
  const parsed = Number.parseInt(params?.[name] ?? defaultValue, 10);
  const value = Number.isFinite(parsed) ? parsed : defaultValue;
  return Math.max(min, Math.min(max, value));
}

module.exports = (params, ctx) => {
  params = params || {};
  const records = intParam(params, "records", 12000, 1000, 50000);
  const rounds = intParam(params, "rounds", 20, 1, 100);
  const payload = new Array(records);
  for (let i = 0; i < records; i++) {
    payload[i] = {
      active: i % 3 !== 0,
      group: i % 97,
      id: i,
      name: `record-${String(i).padStart(6, "0")}`,
      score: (Math.imul(i, 0x9e3779b1) >>> 0) / 4294967296.0,
      tags: [i % 11, i % 17, i % 23],
    };
  }
  let checksum = 0;
  let encodedBytes = 0;
  for (let r = 0; r < rounds; r++) {
    const encoded = JSON.stringify(payload);
    const decoded = JSON.parse(encoded);
    encodedBytes += Buffer.byteLength(encoded, "utf8");
    checksum = (checksum + decoded[records - 1].id + decoded[Math.floor(records / 2)].group) >>> 0;
  }
  return {
    benchmark: "json_dumps",
    records,
    rounds,
    encoded_bytes: encodedBytes,
    checksum,
  };
};
